edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+	}
+	var device v1alpha2.Device
+	var dm v1alpha2.DeviceModel
+	switch resources[1] {
+	case constants.ResourceTypeDevice:
+		switch message.GetOperation() {
+		case model.InsertOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcCreateDevice(&device)
+			if err != nil {
+				klog.Errorf("add device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			deviceList[device.Name] = &device
+		case model.DeleteOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcRemoveDevice(&device)
+			if err != nil {
+				klog.Errorf("delete device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			delete(deviceList, device.Name)
+		case model.UpdateOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcUpdateDevice(&device)
+			if err != nil {
+				klog.Errorf("udpate device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			deviceList[device.Name] = &device
+		default:
+			klog.Warningf("unsupported operation %s", message.GetOperation())
+		}
+	case constants.ResourceTypeDeviceModel:
+		switch message.GetOperation() {
+		case model.InsertOperation:
+			err := json.Unmarshal(message.Content.([]byte), &dm)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcCreateDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("add device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			deviceModelList[dm.Name] = &dm
+		case model.DeleteOperation:
+			err := json.Unmarshal(message.Content.([]byte), &dm)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcRemoveDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("delete device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			delete(deviceModelList, dm.Name)
+		case model.UpdateOperation:
+			err := json.Unmarshal(message.Content.([]byte), &dm)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcUpdateDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("update device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			deviceModelList[dm.Name] = &dm
+		default:
+			klog.Warningf("unsupported operation %s", message.GetOperation())
+		}
+
+	default:
+		klog.Warningf("unsupported resource type %s", resources[3])
+	}
+
+	return nil
+}
+
+func GrpcCreateDevice(device *v1alpha2.Device) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
```

Comment:
these line can be removed

Reply:
i don't think so. i can get your point. if the err is not nil, protocol will be `""` and `GetSockPath` can return `""` and the error can still be handled. but if a mapper registered with `""` protocol, `GetSockPath` may return the sockPath by mistake.